File: playnow-backend/Services/TurfService.cs
using Microsoft.EntityFrameworkCore;
using PlayNow.API.Data;
using PlayNow.API.DTOs;
using PlayNow.API.Models;

namespace PlayNow.API.Services
{
    public interface ITurfService
    {
        Task<List<TurfDetailResponse>> GetAllTurfsAsync();
        Task<TurfDetailResponse> GetTurfByIdAsync(int id);
        Task<TurfDetailResponse> CreateTurfAsync(CreateTurfRequest request);
        Task<TurfDetailResponse> UpdateTurfAsync(int id, UpdateTurfRequest request);
        Task<bool> DeleteTurfAsync(int id);
        Task<List<TurfDetailResponse>> GetTurfsBySportTypeAsync(string sportType);
        Task<List<TurfDetailResponse>> GetTurfsByCityAsync(string city);
    }

    public class TurfService : ITurfService
    {
        private readonly ApplicationDbContext _context;

        public TurfService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<List<TurfDetailResponse>> GetAllTurfsAsync()
        {
            var turfs = await _context.Turfs
                .Include(t => t.TurfAmenities)
                .Include(t => t.TurfTimings)
                .OrderBy(t => t.Name)
                .ToListAsync();

            return turfs.Select(MapToTurfDetailResponse).ToList();
        }

        public async Task<TurfDetailResponse> GetTurfByIdAsync(int id)
        {
            var turf = await _context.Turfs
                .Include(t => t.TurfAmenities)
                .Include(t => t.TurfTimings)
                .FirstOrDefaultAsync(t => t.Id == id);

            if (turf == null)
            {
                throw new InvalidOperationException("Turf not found.");
            }

            return MapToTurfDetailResponse(turf);
        }

        public async Task<TurfDetailResponse> CreateTurfAsync(CreateTurfRequest request)
        {
            var turf = new Turf
            {
                Name = request.Name,
                Description = request.Description,
                Address = request.Address,
                City = request.City,
                PostalCode = request.PostalCode,
                Latitude = request.Latitude,
                Longitude = request.Longitude,
                SportType = request.SportType,
                Capacity = request.Capacity,
                PricePerHour = request.PricePerHour,
                ImageUrl = request.ImageUrl,
                IsAvailable = true,
                IsActive = true,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            _context.Turfs.Add(turf);
            await _context.SaveChangesAsync();

            // Add timings
            if (request.Timings.Any())
            {
                var timings = request.Timings.Select(t => new TurfTiming
                {
                    TurfId = turf.Id,
                    DayOfWeek = t.DayOfWeek,
                    StartTime = t.StartTime,
                    EndTime = t.EndTime,
                    PricePerHour = t.PricePerHour,
                    IsAvailable = t.IsAvailable,
                    CreatedAt = DateTime.UtcNow,
                    UpdatedAt = DateTime.UtcNow
                }).ToList();

                _context.TurfTimings.AddRange(timings);
            }

            // Add amenities
            if (request.Amenities.Any())
            {
                var amenities = request.Amenities.Select(a => new TurfAmenity
                {
                    TurfId = turf.Id,
                    Name = a.Name,
                    Description = a.Description,
                    IsAvailable = a.IsAvailable,
                    AdditionalCost = a.AdditionalCost
                }).ToList();

                _context.TurfAmenities.AddRange(amenities);
            }

            await _context.SaveChangesAsync();

            return await GetTurfByIdAsync(turf.Id);
        }

        public async Task<TurfDetailResponse> UpdateTurfAsync(int id, UpdateTurfRequest request)
        {
            var turf = await _context.Turfs
                .Include(t => t.TurfAmenities)
                .Include(t => t.TurfTimings)
                .FirstOrDefaultAsync(t => t.Id == id);

            if (turf == null)
            {
                throw new InvalidOperationException("Turf not found.");
            }

            // Update turf properties
            turf.Name = request.Name;
            turf.Description = request.Description;
            turf.Address = request.Address;
            turf.City = request.City;
            turf.PostalCode = request.PostalCode;
            turf.Latitude = request.Latitude;
            turf.Longitude = request.Longitude;
            turf.SportType = request.SportType;
            turf.Capacity = request.Capacity;
            turf.PricePerHour = request.PricePerHour;
            turf.ImageUrl = request.ImageUrl;
            turf.IsAvailable = request.IsAvailable;
            turf.IsActive = request.IsActive;
            turf.UpdatedAt = DateTime.UtcNow;

            // Update timings
            _context.TurfTimings.RemoveRange(turf.TurfTimings);
            if (request.Timings.Any())
            {
                var timings = request.Timings.Select(t => new TurfTiming
                {
                    TurfId = turf.Id,
                    DayOfWeek = t.DayOfWeek,
                    StartTime = t.StartTime,
                    EndTime = t.EndTime,
                    PricePerHour = t.PricePerHour,
                    IsAvailable = t.IsAvailable,
                    CreatedAt = DateTime.UtcNow,
                    UpdatedAt = DateTime.UtcNow
                }).ToList();

                _context.TurfTimings.AddRange(timings);
            }

            // Update amenities
            _context.TurfAmenities.RemoveRange(turf.TurfAmenities);
            if (request.Amenities.Any())
            {
                var amenities = request.Amenities.Select(a => new TurfAmenity
                {
                    TurfId = turf.Id,
                    Name = a.Name,
                    Description = a.Description,
                    IsAvailable = a.IsAvailable,
                    AdditionalCost = a.AdditionalCost
                }).ToList();

                _context.TurfAmenities.AddRange(amenities);
            }

            await _context.SaveChangesAsync();

            return await GetTurfByIdAsync(turf.Id);
        }

        public async Task<bool> DeleteTurfAsync(int id)
        {
            var turf = await _context.Turfs.FindAsync(id);
            if (turf == null)
            {
                return false;
            }

            _context.Turfs.Remove(turf);
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<List<TurfDetailResponse>> GetTurfsBySportTypeAsync(string sportType)
        {
            var turfs = await _context.Turfs
                .Include(t => t.TurfAmenities)
                .Include(t => t.TurfTimings)
                .Where(t => t.SportType == sportType && t.IsActive)
                .OrderBy(t => t.Name)
                .ToListAsync();

            return turfs.Select(MapToTurfDetailResponse).ToList();
        }

        public async Task<List<TurfDetailResponse>> GetTurfsByCityAsync(string city)
        {
            var turfs = await _context.Turfs
                .Include(t => t.TurfAmenities)
                .Include(t => t.TurfTimings)
                .Where(t => t.City == city && t.IsActive)
                .OrderBy(t => t.Name)
                .ToListAsync();

            return turfs.Select(MapToTurfDetailResponse).ToList();
        }

        private static TurfDetailResponse MapToTurfDetailResponse(Turf turf)
        {
            return new TurfDetailResponse
            {
                Id = turf.Id,
                Name = turf.Name,
                Description = turf.Description,
                Address = turf.Address,
                City = turf.City,
                PostalCode = turf.PostalCode,
                Latitude = turf.Latitude,
                Longitude = turf.Longitude,
                SportType = turf.SportType,
                Capacity = turf.Capacity,
                PricePerHour = turf.PricePerHour,
                ImageUrl = turf.ImageUrl,
                IsAvailable = turf.IsAvailable,
                IsActive = turf.IsActive,
                CreatedAt = turf.CreatedAt,
                UpdatedAt = turf.UpdatedAt,
                Timings = turf.TurfTimings?.Select(t => new TurfTimingResponse
                {
                    Id = t.Id,
                    DayOfWeek = t.DayOfWeek,
                    StartTime = t.StartTime,
                    EndTime = t.EndTime,
                    PricePerHour = t.PricePerHour,
                    IsAvailable = t.IsAvailable
                }).ToList() ?? new List<TurfTimingResponse>(),
                Amenities = turf.TurfAmenities?.Select(a => new TurfAmenityResponse
                {
                    Id = a.Id,
                    Name = a.Name,
                    Description = a.Description,
                    IsAvailable = a.IsAvailable,
                    AdditionalCost = a.AdditionalCost
                }).ToList() ?? new List<TurfAmenityResponse>()
            };
        }
    }
} 

File: playnow-frontend/src/app/core/services/turf.service.ts
import { Injectable, signal, computed } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap, catchError, throwError, map } from 'rxjs';

export interface TurfDetail {
  id: number;
  name: string;
  description: string;
  address: string;
  city: string;
  postalCode: string;
  latitude: number;
  longitude: number;
  sportType: string;
  capacity: number;
  pricePerHour: number;
  imageUrl?: string;
  isAvailable: boolean;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  timings: TurfTiming[];
  amenities: TurfAmenity[];
}

export interface TurfTiming {
  id: number;
  dayOfWeek: string;
  startTime: string;
  endTime: string;
  pricePerHour: number;
  isAvailable: boolean;
}

export interface TurfAmenity {
  id: number;
  name: string;
  description?: string;
  isAvailable: boolean;
  additionalCost?: number;
}

export interface CreateTurfRequest {
  name: string;
  description: string;
  address: string;
  city: string;
  postalCode: string;
  latitude: number;
  longitude: number;
  sportType: string;
  capacity: number;
  pricePerHour: number;
  imageUrl?: string;
  timings: TurfTimingRequest[];
  amenities: TurfAmenityRequest[];
}

export interface UpdateTurfRequest extends CreateTurfRequest {
  isAvailable: boolean;
  isActive: boolean;
}

export interface TurfTimingRequest {
  dayOfWeek: string;
  startTime: string;
  endTime: string;
  pricePerHour: number;
  isAvailable: boolean;
}

export interface TurfAmenityRequest {
  name: string;
  description?: string;
  isAvailable: boolean;
  additionalCost?: number;
}

@Injectable({
  providedIn: 'root'
})
export class TurfService {
  private turfs = signal<TurfDetail[]>([]);
  private isLoading = signal(false);
  private error = signal<string | null>(null);

  private readonly API_BASE_URL = 'http://localhost:5048/api';

  // Computed signals
  public readonly allTurfs = this.turfs.asReadonly();
  public readonly isLoadingState = this.isLoading.asReadonly();
  public readonly errorState = this.error.asReadonly();

  public readonly availableTurfs = computed(() => 
    this.turfs().filter(turf => turf.isAvailable && turf.isActive)
  );

  public readonly turfsBySport = computed(() => {
    const turfs = this.turfs();
    const grouped: { [key: string]: TurfDetail[] } = {};
    turfs.forEach(turf => {
      if (!grouped[turf.sportType]) {
        grouped[turf.sportType] = [];
      }
      grouped[turf.sportType].push(turf);
    });
    return grouped;
  });

  constructor(private http: HttpClient) {
    this.loadTurfs();
  }

  /**
   * Load all turfs from backend
   */
  loadTurfs(): Observable<TurfDetail[]> {
    this.isLoading.set(true);
    this.error.set(null);

    return this.http.get<TurfDetail[]>(`${this.API_BASE_URL}/turf`)
      .pipe(
        tap(turfs => {
          this.turfs.set(turfs);
          this.isLoading.set(false);
        }),
        catchError(error => {
          this.isLoading.set(false);
          this.error.set('Failed to load turfs');
          return throwError(() => error);
        })
      );
  }

  /**
   * Get turf by ID
   */
  getTurfById(id: number): Observable<TurfDetail> {
    return this.http.get<TurfDetail>(`${this.API_BASE_URL}/turf/${id}`)
      .pipe(
        catchError(error => {
          this.error.set('Failed to load turf details');
          return throwError(() => error);
        })
      );
  }

  /**
   * Create new turf
   */
  createTurf(request: CreateTurfRequest): Observable<TurfDetail> {
    this.isLoading.set(true);
    this.error.set(null);

    return this.http.post<TurfDetail>(`${this.API_BASE_URL}/turf`, request)
      .pipe(
        tap(turf => {
          this.turfs.update(turfs => [...turfs, turf]);
          this.isLoading.set(false);
        }),
        catchError(error => {
          this.isLoading.set(false);
          this.error.set('Failed to create turf');
          return throwError(() => error);
        })
      );
  }

  /**
   * Update existing turf
   */
  updateTurf(id: number, request: UpdateTurfRequest): Observable<TurfDetail> {
    this.isLoading.set(true);
    this.error.set(null);

    return this.http.put<TurfDetail>(`${this.API_BASE_URL}/turf/${id}`, request)
      .pipe(
        tap(updatedTurf => {
          this.turfs.update(turfs => 
            turfs.map(turf => turf.id === id ? updatedTurf : turf)
          );
          this.isLoading.set(false);
        }),
        catchError(error => {
          this.isLoading.set(false);
          this.error.set('Failed to update turf');
          return throwError(() => error);
        })
      );
  }

  /**
   * Delete turf
   */
  deleteTurf(id: number): Observable<boolean> {
    this.isLoading.set(true);
    this.error.set(null);

    return this.http.delete(`${this.API_BASE_URL}/turf/${id}`)
      .pipe(
        tap(() => {
          this.turfs.update(turfs => turfs.filter(turf => turf.id !== id));
          this.isLoading.set(false);
        }),
        catchError(error => {
          this.isLoading.set(false);
          this.error.set('Failed to delete turf');
          return throwError(() => error);
        }),
        map(() => true)
      );
  }

  /**
   * Get turfs by sport type
   */
  getTurfsBySportType(sportType: string): Observable<TurfDetail[]> {
    return this.http.get<TurfDetail[]>(`${this.API_BASE_URL}/turf/sport/${sportType}`)
      .pipe(
        catchError(error => {
          this.error.set('Failed to load turfs by sport type');
          return throwError(() => error);
        })
      );
  }

  /**
   * Get turfs by city
   */
  getTurfsByCity(city: string): Observable<TurfDetail[]> {
    return this.http.get<TurfDetail[]>(`${this.API_BASE_URL}/turf/city/${city}`)
      .pipe(
        catchError(error => {
          this.error.set('Failed to load turfs by city');
          return throwError(() => error);
        })
      );
  }

  /**
   * Get turf from local state by ID
   */
  getTurfFromState(id: number): TurfDetail | undefined {
    return this.turfs().find(turf => turf.id === id);
  }

  /**
   * Clear error state
   */
  clearError(): void {
    this.error.set(null);
  }

  /**
   * Get available time slots for a specific date
   */
  getAvailableSlots(turf: TurfDetail, date: string): string[] {
    const selectedDay = new Date(date).toLocaleDateString('en-US', { weekday: 'long' });
    const dayTiming = turf.timings.find(t => t.dayOfWeek === selectedDay);

    if (!dayTiming || !dayTiming.isAvailable) {
      return [];
    }

    const slots: string[] = [];
    const startTime = new Date(`2000-01-01T${dayTiming.startTime}`);
    const endTime = new Date(`2000-01-01T${dayTiming.endTime}`);
    
    // Generate 2-hour slots
    let currentTime = new Date(startTime);
    while (currentTime < endTime) {
      const slotEnd = new Date(currentTime.getTime() + 2 * 60 * 60 * 1000); // 2 hours
      if (slotEnd <= endTime) {
        slots.push(currentTime.toTimeString().slice(0, 5));
      }
      currentTime = slotEnd;
    }

    return slots;
  }

  /**
   * Calculate total price for booking
   */
  calculateBookingPrice(turf: TurfDetail, startTime: string, endTime: string): number {
    const start = new Date(`2000-01-01T${startTime}`);
    const end = new Date(`2000-01-01T${endTime}`);
    const duration = Math.ceil((end.getTime() - start.getTime()) / (1000 * 60 * 60));
    
    const basePrice = turf.pricePerHour * duration;
    
    // Add amenity costs
    let amenityCost = 0;
    turf.amenities.forEach(amenity => {
      if (amenity.isAvailable && amenity.additionalCost) {
        amenityCost += amenity.additionalCost;
      }
    });
    
    return basePrice + amenityCost;
  }
} 